### PR TITLE
XLIFF Sync - New Build with Translations command

### PR DESCRIPTION
**Pull request by FlorianNoeverGOB**
### Build with Translations
This adds a new command `xliffSync.buildWithTranslations` that combines functionalities to:
> 1. Delete current translation files
> 2. Build App using the build command defined in new Setting `Build Command To Execute`
> 3. Create new Target files for Languages defined in new Setting `Default Languages`
> 4. Syncs Translations files to show if translations are missing

Therefore 2 new settings are created:

#### Default Languages
Xliff file generation now uses the languages defined in this Setting
Default Languages Setting
When the setting is empty (default) the user is asked to specifiy the languages (as normally)

### Build Command To Execute
When using the new Command `Build with Translations` the command specified in this Setting will be used for building the App
Build Command To Execute Setting

</br>

*There are multiple whitespace changes due to VSCode handling the formatting.*

if there are any questions or anything else feel free to contact me :)

### CODE REVIEW

**Comment by rvanbekkum** [comment]
Thank you. This indeed looks very useful! 😊

I will check the PR more thoroughly, but I do already have some small suggestions for changes before we merge this into the develop branch.

**Comment by rvanbekkum** [comment]
> I can't seem to find any comments. But I don't use GitHub that much so perhaps I missed them. I looked in Files changed where I can make comments to files (lines) but there wasn't any comment. Where did you put them? 😅

Apparently I needed to click on a _Submit Review_ button before the comments would be visible. 😅
(Shows that I am more used to Azure DevOps where the comments are posted directly 🙊)

You should be able to see them now.

File: package.json [hunk 1]
```diff
@@ -383,6 +384,28 @@
                     "default": "TargetLanguageCode",
                     "description": "Specifies which target language to use by default in the XLIFF Sync snippets.",
                     "scope": "resource"
+                },
+                "xliffSync.defaultLanguages": {
```
[the remark this answers is not in the document]
Comment (FlorianNoeverGOB) [reply package.json:388]: I wanted to leave this empty because `createNewTargetFiles` also uses this setting and when empty asks for the languages to use as it would do normally.

**Comment by FlorianNoeverGOB** [comment]
Applied all comments. Take your time checking this, there is no hurry 😄 . I reacted with "👍" on every comment I "solved" if you agree with me you can "resolve" the comment (I would say I don't resolve myself cause maybe you wont agree with my "solving" so I just react)

File: README.md [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # XLIFF Sync
 
-[![License: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT) [![Version](https://vsmarketplacebadges.dev/version-short/rvanbekkum.xliff-sync.svg?color=blue)](https://marketplace.visualstudio.com/items?itemName=rvanbekkum.xliff-sync) [![Installs](https://vsmarketplacebadges.dev/installs/rvanbekkum.xliff-sync.svg?color=blue)](https://marketplace.visualstudio.com/items?itemName=rvanbekkum.xliff-sync)
+[![License: MIT](https://img.shields.io/badge/License-MIT-blue.png)](https://opensource.org/licenses/MIT)
```
Comment (rvanbekkum) [inline README.md:3]: Any particular reason for changing this in the README?

Comment (rvanbekkum) [reply README.md:3]: i.e. changing it from svg to png and removing the other properties?

Comment (FlorianNoeverGOB) [reply README.md:3]: ah yes it wouldn't let me create the vsix package otherwise (doesnt support svg in readme) at least tool I used to build (vsce). But as this wasn't a problem for you I just revert the changes.

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (rvanbekkum) [reply package.json:387]: But you mean that's also when explicitly defining the default to be empty?

[the remark this answers is not in the document]
Comment (FlorianNoeverGOB) [reply package.json:387]: Already set the default to empty 👍

**Review by rvanbekkum — COMMENTED** [review]
Just took some more time to better review your PR contentwise.

I was wondering why the "Build with Translations" command is deleting existing translation files?
I would prefer it to reuse the existing translation files, so that you don't lose your existing translations. I don't see the need to remove them and I think it might work counterproductive.

Could you check and/or explain?

File: resources/custom_xliffSync_buildCommandToExecute.png
Comment (rvanbekkum) [inline resources/custom_xliffSync_buildCommandToExecute.png:1]: Rename (remove the `custom_`).
Also for the other new files.

Comment (rvanbekkum) [reply resources/custom_xliffSync_buildCommandToExecute.png:1]: Now that I think of it, these files are probably not referenced anymore now that the 'custom README' was removed.

File: resources/custom_xliffSync_installSteps.png
Comment (rvanbekkum) [inline resources/custom_xliffSync_installSteps.png:1]: This one can be removed.

File: src/features/tools/files-helper.ts [hunk 3]
```diff
@@ -51,30 +51,48 @@ export class FilesHelper {
...
 
     return uris;
   }
 
+  /**
+    * Checks if translation files exist in the opened workspace
+    *
+    * @param {WorkspaceFolder} workspaceFolder The folder to restrict the search to.
+    *
+    * @returns A boolean that specifies wether translation files exist in the current workspace.
```
Comment (rvanbekkum) [inline src/features/tools/files-helper.ts:83]: wether -> whether

File: src/features/tools/files-helper.ts [hunk 4]
```diff
@@ -51,30 +51,48 @@ export class FilesHelper {
...
+    */
+  public static async checkXliffFilesExist(workspaceFolder?: WorkspaceFolder): Promise<boolean> {
```
Comment (rvanbekkum) [inline src/features/tools/files-helper.ts:85]: Move to FilesHelper?

File: src/features/trans-sync.ts [hunk 5]
```diff
@@ -381,3 +444,13 @@ async function autoRunTranslationChecks(workspaceFolder?: WorkspaceFolder, targe
 
     await runTranslationChecksForWorkspaceFolder(autoCheckMissingTranslations, autoCheckNeedWorkTranslations, targetUri, workspaceFolder);
 }
+
+function getCurrentWorkspaceFolder(): WorkspaceFolder | undefined {
```
Comment (rvanbekkum) [inline src/features/trans-sync.ts:448]: Use method from FilesHelper for selection of workspace folder?

File: package.json [hunk 6]
```diff
@@ -383,6 +383,29 @@
...
+                    "type": "array",
+                    "items": {
+                        "type": "string",
+                        "pattern": "^[a-z]{2}-[A-Z]{2}$"
```
[the remark this answers is not in the document]
Comment (FlorianNoeverGOB) [reply package.json:391]: Is there a pattern that works for all? Or is it the best to just remove the pattern entirely?

**Comment by FlorianNoeverGOB** [comment]
> Just took some more time to better review your PR contentwise.
> 
> I was wondering why the "Build with Translations" command is deleting existing translation files? I would prefer it to reuse the existing translation files, so that you don't lose your existing translations. I don't see the need to remove them and I think it might work counterproductive.
> 
> Could you check and/or explain?

So basically we don't use the Sync Feature because only our programmers create the Translations. So we don't want to have the "needs-work" (or how it was called) tags, because we always use the translations we created in the files. So if for example I add a new Label with a translation or change the translation of a label, that new translation will only get reviewed in the associated Pull Request by the Reviewer and then go directly into the Customers System. If I were to use the Sync command after I changed the Translation, I would only get the "needs-work" Tag added in the xliff files but not the new Translation (right? 😅). But as the new Translation is directly also the "final form" it should get added to the xliff files as normal which is why we normally delete the xliff files and call "create new target files" again.

I hope that was understandable 😅 if something was unclear please let me know 😄

**Comment by rvanbekkum** [comment]
> So basically we don't use the Sync Feature because only our programmers create the Translations. So we don't want to have the "needs-work" (or how it was called) tags, because we always use the translations we created in the files. So if for example I add a new Label with a translation or change the translation of a label, that new translation will only get reviewed in the associated Pull Request by the Reviewer and then go directly into the Customers System. If I were to use the Sync command after I changed the Translation, I would only get the "needs-work" Tag added in the xliff files but not the new Translation (right? 😅). But as the new Translation is directly also the "final form" it should get added to the xliff files as normal which is why we normally delete the xliff files and call "create new target files" again.
> 
> I hope that was understandable 😅 if something was unclear please let me know 😄

Do you mean you use the `parseFromDeveloperNote` functionality?
If so, then that has some configuration options in place for when you would like to have it overwrite translations (`xliffSync.parseFromDeveloperNoteOverwrite`) and/or ignore source text changes (`xliffSync.detectSourceTextChanges`).
If you see any problems with using these settings, then please let me know as well.

Considering there are also companies that do _not_ use the parseFromDeveloperNote functionality (our company included), I think it would be better to have this "Build with Translations" do a "build + sync" instead of a "build + delete + recreate" so that you do not lose the translations you had in the target translation .xlf files.

Is that something you could change?

File: package.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (rvanbekkum) [reply package.json:391]: They are RFC 4646 or RFC 5646 language tags, but I think it might be better to just remove the pattern as I otherwise a RegEx that is a bit too complex. I think it's good enough if people can type in anything they want and think for themselves whether it is a valid tag. ;)

**Comment by FlorianNoeverGOB** [comment]
> > So basically we don't use the Sync Feature because only our programmers create the Translations. So we don't want to have the "needs-work" (or how it was called) tags, because we always use the translations we created in the files. So if for example I add a new Label with a translation or change the translation of a label, that new translation will only get reviewed in the associated Pull Request by the Reviewer and then go directly into the Customers System. If I were to use the Sync command after I changed the Translation, I would only get the "needs-work" Tag added in the xliff files but not the new Translation (right? 😅). But as the new Translation is directly also the "final form" it should get added to the xliff files as normal which is why we normally delete the xliff files and call "create new target files" again.
> > I hope that was understandable 😅 if something was unclear please let me know 😄
> 
> Do you mean you use the `parseFromDeveloperNote` functionality? If so, then that has some configuration options in place for when you would like to have it overwrite translations (`xliffSync.parseFromDeveloperNoteOverwrite`) and/or ignore source text changes (`xliffSync.detectSourceTextChanges`). If you see any problems with using these settings, then please let me know as well.
> 
> Considering there are also companies that do _not_ use the parseFromDeveloperNote functionality (our company included), I think it would be better to have this "Build with Translations" do a "build + sync" instead of a "build + delete + recreate" so that you do not lose the translations you had in the target translation .xlf files.
> 
> Is that something you could change?

Yes we use `parseFromDeveloperNote`. Actually these are our settings:
image

I am unsure whether a Build + Sync would work. It should but I remember there was a case where it wouldn't work which is why we always do this for updating the translations in our company:
Build -> Delete Target Translation Files -> `createNewTargetFiles`

Im going to Test a bit whether I can find that case again where it wouldn't work. There is also the possibility of having both in 2 commands: `Build with Translations (Sync)` and `Build with Translations (Recreate)`

**Comment by FlorianNoeverGOB** [comment]
> > > So basically we don't use the Sync Feature because only our programmers create the Translations. So we don't want to have the "needs-work" (or how it was called) tags, because we always use the translations we created in the files. So if for example I add a new Label with a translation or change the translation of a label, that new translation will only get reviewed in the associated Pull Request by the Reviewer and then go directly into the Customers System. If I were to use the Sync command after I changed the Translation, I would only get the "needs-work" Tag added in the xliff files but not the new Translation (right? 😅). But as the new Translation is directly also the "final form" it should get added to the xliff files as normal which is why we normally delete the xliff files and call "create new target files" again.
> > > I hope that was understandable 😅 if something was unclear please let me know 😄
> > 
> > 
> > Do you mean you use the `parseFromDeveloperNote` functionality? If so, then that has some configuration options in place for when you would like to have it overwrite translations (`xliffSync.parseFromDeveloperNoteOverwrite`) and/or ignore source text changes (`xliffSync.detectSourceTextChanges`). If you see any problems with using these settings, then please let me know as well.
> > Considering there are also companies that do _not_ use the parseFromDeveloperNote functionality (our company included), I think it would be better to have this "Build with Translations" do a "build + sync" instead of a "build + delete + recreate" so that you do not lose the translations you had in the target translation .xlf files.
> > Is that something you could change?
> 
> Yes we use `parseFromDeveloperNote`. Actually these are our settings: image
> 
> I am unsure whether a Build + Sync would work. It should but I remember there was a case where it wouldn't work which is why we always do this for updating the translations in our company: Build -> Delete Target Translation Files -> `createNewTargetFiles`
> 
> Im going to Test a bit whether I can find that case again where it wouldn't work. There is also the possibility of having both in 2 commands: `Build with Translations (Sync)` and `Build with Translations (Recreate)`

Ok so lets say we have a Project with some custom Code and ofcourse Labels and Translations for them. The Customer now wants to change the tooltip of a field. Now I go and add whatever the request was as the new Tooltip and its translation.
I build the using `al.package` (ctrl+shift+b)

`SyncTranslationUnits`:
If I now execute the sync command. The field gets a new state `needs-adaptation`  for the translation. Because the source text (the original english tooltip) changed. And now I don't have the new tooltip and its translation but still the old one.
If I would only change the translation of the tooltip, this would work fine. But If I also change the original source text we have this problem.

`CreateNewTargetFiles`:
If I instead delete the translation files and create them with the create command. The new tooltip is there with its translation. So now I can PR my changes into the main branch and make an update for the Customers System so he has the new tooltip

So basically I could either provide both variants with my build command like I said in my last message.
Or we would need an option for ignoring source text changes and just automatically translate them without getting the 'needs-adaptation' state.

**Comment by rvanbekkum** [comment]
> Or we would need an option for ignoring source text changes and just automatically translate them without getting the 'needs-adaptation' state.

That already exists, i.e., `xliffSync.detectSourceTextChanges`. See my earlier message.
So if you use that setting then you should be good to go. 😉

**Comment by rvanbekkum** [comment]
I am not sure what you mean. The base translation file (.g.xlf) is what the AL compiler generates and contains the source text in English (United States) en-US. (Strictly seen, therefore there is also not really a need for a separate .en-US.xlf file)
XLIFF Sync just synchronizes to all other (target) .xlf files that you may have in your project.

If you would like me to have a look maybe you can share a sample project/xliff files to reproduce the problem, together with the VS Code settings you are using. You could also file that as a GitHub issue in this repository.

**Comment by FlorianNoeverGOB** [comment]
We always use en-US and (mostly) de-DE as our languages. So 3 files:
.d.xlf
en-US.xlf
de-DE.xlf

(en-US so you could in theory sent it someone ... i guess)
 And the problem I mentioned only appeard on the en-US.xlf file. It seems to be just that. Use languages en-US and some other. change some translation and source text (while using detectSource: false and parseDeloperNoteOverwrite: true) and build + sync. en-US now still has `needs-adaptation` and old translation (old source text). But if you want I can create a sample project. 

I will also make an Issue to this.

**Comment by FlorianNoeverGOB** [comment]
Everything works now. I will modify the command to just build + sync now

**Comment by rvanbekkum** [comment]
Hi @FlorianNoeverGOB,

Thanks for the changes. I have made some last changes to the PR and pushed a commit. I have replaced the statements that invoked the synchronization with a call to an existing method for it and I made some updates to the CHANGELOG and README.
Could you please check these changes before I go over and merge them?

I have tested the functionality some further and it seems to work great! 😊

**Comment by FlorianNoeverGOB** [comment]
> Hi @FlorianNoeverGOB,
> 
> Thanks for the changes. I have made some last changes to the PR and pushed a commit. I have replaced the statements that invoked the synchronization with a call to an existing method for it and I made some updates to the CHANGELOG and README. Could you please check these changes before I go over and merge them?
> 
> I have tested the functionality some further and it seems to work great! 😊

Also tested it and seems to be working fine. So you can go ahead and merge 👍
